unittests/runtime-io/io.cpp

```diff
@@ -0,0 +1,343 @@
+#include<boost/test/unit_test.hpp>
+#include<gmp.h>
+#include<mpfr.h>
+#include<cstdint>
+#include<cstdlib>
+#include<cstring>
+
+#include "runtime/header.h"
+#include "runtime/alloc.h"
+
+#include "fcntl.h"
+#include "unistd.h"
+
+#define KCHAR char
+extern "C" {
+
+#define ERRTAG(err) "Lbl'hash'" #err "{}"
+#define ERRBLOCK(tag) (((uint64_t)tag << 32) | 1)
+#define NUM_SYMBOLS 5
+  const char * symbols[NUM_SYMBOLS] = {ERRTAG(EOF), ERRTAG(ENOENT), ERRTAG(EBADF), ERRTAG(EOVERFLOW), ERRTAG(EACCES)};
+
+  uint32_t getTagForSymbolName(const char *s) {
+    for (int i = 0; i < NUM_SYMBOLS; i++) {
+      if (0 == strcmp(symbols[i], s)) {
+        return i;
+      }
+    }
+
+    return 0;
+  }
+  
+  struct blockheader getBlockHeaderForSymbol(uint32_t tag) {
+    return blockheader {0};
+  }
+
+  void add_hash64(void*, uint64_t) {}
+
+  void flush_IO_logs();
+  string * makeString(const KCHAR *, int64_t len = -1);
+  blockheader header_err();
+  block * hook_IO_open(string * filename, string * control);
+  block * hook_IO_tell(mpz_t i);
+  block * hook_IO_getc(mpz_t i);
+  block * hook_IO_read(mpz_t i, mpz_t len);
+  block * hook_IO_close(mpz_t i);
+  block * hook_IO_seek(mpz_t i, mpz_t loc);
+  block * hook_IO_seekEnd(mpz_t i, mpz_t loc);
+  block * hook_IO_putc(mpz_t i, mpz_t c);
+  block * hook_IO_write(mpz_t i, string * str);
+  block * hook_IO_lock(mpz_t i, mpz_t len);
+  block * hook_IO_unlock(mpz_t i, mpz_t len);
+  block * hook_IO_log(string * path, string * msg);
+
+  mpz_ptr move_int(mpz_t i) {
+    mpz_ptr result = (mpz_ptr)malloc(sizeof(__mpz_struct));
+    *result = *i;
+    return result;
+  }
+
+  floating *move_float(floating *i) {
+    floating *result = (floating *)malloc(sizeof(floating));
+    *result = *i;
+    return result;
+  }
+}
+
+int overwriteTestFile() {
+  int fd = ::open("test.txt", O_RDWR | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH);
+  ::write(fd, "hello world!", 12);
+  ::lseek(fd, 0, SEEK_SET);
+  return fd;
+}
+
+BOOST_AUTO_TEST_SUITE(IoTest)
+
+BOOST_AUTO_TEST_CASE(open) {
+  int fd = overwriteTestFile();
+  ::close(fd);
+
+  auto realFilename = makeString("test.txt");
+  auto fakeFilename = makeString("testFake.txt");
+  auto control = makeString("r");
+
+  block * b1 = hook_IO_open(realFilename, control);
+  BOOST_CHECK(0 < mpz_cmp_si((mpz_ptr) *(b1->children), 0));
```

Comment:
I'm concerned that this test will probably invoke undefined behavior if a bug causes open to not return an integer. It would be nice to explicitly assign mocked headers for inj{IOError} and inj{Int} and check against them in the test, just to make sure.

Reply:
We should do this for all the tests, too.